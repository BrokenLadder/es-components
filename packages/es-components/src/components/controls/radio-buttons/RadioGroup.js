import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';

import Fieldset from '../../containers/fieldset/Fieldset';

import RadioButton from './RadioButton';

function RadioGroup({
  name,
  radioOptions,
  legendContent,
  value,
  hasError = false,
  disableAllOptions = false,
  inline = true,
  onChange = noop,
  extraContent
}) {
  return (
    <Fieldset legendContent={legendContent}>
      {extraContent}
      {radioOptions.map((config, index) => {
        const radioId = `${name}-option-${index + 1}`;
        const isChecked = config.optionValue === value;
        const isDisabled = disableAllOptions || config.isDisabled;
        const radioProps = {
          name,
          isChecked,
          hasError,
          isDisabled,
          inline,
          onChange,
          id: radioId,
          optionText: config.optionText,
          value: config.optionValue
        };
        return <RadioButton key={radioId} {...radioProps} />;
      })}
    </Fieldset>
  );
}

const radioOptionShape = {
  optionText: PropTypes.string.isRequired,
  optionValue: PropTypes.any.isRequired,
  /** Render this option as disabled */
  isDisabled: PropTypes.bool
};

RadioGroup.propTypes = {
  /** The name of the radio group */
  name: PropTypes.string.isRequired,
  /** The content to render in the legend */
  legendContent: PropTypes.node,
  /** Options for the radio group to display */
  radioOptions: PropTypes.arrayOf(PropTypes.shape(radioOptionShape)).isRequired,
  /** Selected option for the radio group */
  value: PropTypes.any,
  /** Display all radio buttons in an errored state */
  hasError: PropTypes.bool,
  /** Disable all radio buttons */
  disableAllOptions: PropTypes.bool,
  /** Display the radio buttons inline */
  inline: PropTypes.bool,
  /** Function to execute when a radio button is selected */
  onChange: PropTypes.func,
  /** Extra content that can be rendered after the Legend but before the radio buttons, allows
   * content to be put in that will not affect the accessability of the Legend/Radio button relationship.
   */
  extraContent: PropTypes.node
};

RadioGroup.defaultProps = {
  legendContent: null,
  value: null,
  hasError: false,
  disableAllOptions: false,
  inline: false,
  onChange: noop,
  extraContent: null
};

export default RadioGroup;
