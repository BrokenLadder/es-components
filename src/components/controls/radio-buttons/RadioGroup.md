Group radio buttons together using a ``RadioGroup`` component.

```
const options = [{
  optionText: 'Red',
  value: 'red'
}, {
  optionText: 'Green',
  value: 'green'
}, {
  optionText: 'Blue',
  value: 'blue'
}];

<RadioGroup name="colors" radioOptions={options} checkedValue="green" />
```

Providing a ``legendText`` option will render a legend with the grouped radio buttons.

```
const options = [{
  optionText: 'Hiking',
  value: 'hiking'
}, {
  optionText: 'Biking',
  value: 'biking'
}, {
  optionText: 'kayaking',
  value: 'kayaking'
}];

<RadioGroup
  name="recreational-activities"
  legendText="Recreational activities"
  radioOptions={options}
/>
```

Setting the ``inline`` option to false will stack the radio buttons

```
const options = [{
  optionText: 'Meat',
  value: 'meat'
}, {
  optionText: 'Seafood',
  value: 'seafood'
}, {
  optionText: 'Vegetarian',
  value: 'vegetarian'
}];

<RadioGroup
  name="plates"
  radioOptions={options}
  inline={false}
/>
```

Each radio is displayed as an error when the ``hasError`` prop is true. An errored radio group with a default checked option is rendered with a filled radio button.

```
const options = [{
  optionText: 'First',
  value: '1'
}, {
  optionText: 'Second',
  value: '2'
}, {
  optionText: 'Third',

value: '3'
}];

<RadioGroup
  name="placings"
  radioOptions={options}
  hasError
/>
```

```
const options = [{
  optionText: 'Planes',
  value: 'plane'
}, {
  optionText: 'Trains',
  value: 'train'
}, {
  optionText: 'Automobiles',
  value: 'automobile'
}];

<RadioGroup
  name="transports"
  radioOptions={options}
  checkedValue="plane"
  hasError
/>
```

Each radio is disabled when the ``disableAllOptions`` prop is true. A disabled radio group with a default checked option is rendered with a filled radio button.

```
const options = [{
  optionText: 'Apple',
  value: 'apple'
}, {
  optionText: 'Banana',
  value: 'banana'
}, {
  optionText: 'Pear',
  value: 'pear'
}];

<RadioGroup
  name="fruits"
  radioOptions={options}
  disableAllOptions
/>
```

```
const options = [{
  optionText: 'Books',
  value: 'books'
}, {
  optionText: 'Television',
  value: 'tv'
}];

<RadioGroup
  name="media-preference"
  radioOptions={options}
  disableAllOptions
  checkedValue="books"
/>
```

Individual options can also be disabled.
```
const options = [{
  optionText: 'Hamburger',
  value: 'burger'
}, {
  optionText: 'Cheeseburger',
  value: 'cheeseburger'
}, {
  optionText: 'Hotdog',
  value: 'hotdog',
  disabled: true
}];

<RadioGroup
  name="cookout-food"
  radioOptions={options}
/>
```
