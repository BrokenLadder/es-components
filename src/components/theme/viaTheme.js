const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
