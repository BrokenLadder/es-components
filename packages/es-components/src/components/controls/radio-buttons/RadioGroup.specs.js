/* eslint-env jest */

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { range } from 'lodash';
import viaTheme from 'es-components-via-theme';
import { ThemeProvider } from 'styled-components';

import RadioGroup from './RadioGroup';

function buildOptions(numberOfOptions, optionIndexToDisable) {
  return range(0, numberOfOptions).map(idx => ({
    optionText: `Option ${idx}`,
    optionValue: idx,
    disabled: idx === optionIndexToDisable
  }));
}

beforeEach(cleanup);

it('renders each radio input as disabled when disableAllOptions is true', () => {
  const options = buildOptions(3);
  const { container } = render(
    <ThemeProvider theme={viaTheme}>
      <RadioGroup name="test" radioOptions={options} disableAllOptions />
    </ThemeProvider>
  );

  const inputs = container.querySelectorAll('input');

  expect(inputs[0]).toBeDisabled();
  expect(inputs[1]).toBeDisabled();
  expect(inputs[2]).toBeDisabled();
});

it('renders a specific radio input as disabled when that option is set to disabled', () => {
  const options = buildOptions(3, 0);
  const { container } = render(
    <ThemeProvider theme={viaTheme}>
      <RadioGroup name="test" radioOptions={options} />
    </ThemeProvider>
  );

  const inputs = container.querySelectorAll('input');

  expect(inputs[0]).toBeDisabled();
  expect(inputs[1]).not.toBeDisabled();
  expect(inputs[2]).not.toBeDisabled();
});

it('renders legend when provided', () => {
  const { queryByText } = render(
    <ThemeProvider theme={viaTheme}>
      <RadioGroup
        name="test"
        radioOptions={buildOptions(3)}
        legendContent="Test legend"
      />
    </ThemeProvider>
  );

  expect(queryByText('Test legend')).not.toBeNull();
});

it('renders intro when provided', () => {
  const { queryByText } = render(
    <ThemeProvider theme={viaTheme}>
      <RadioGroup
        name="test"
        radioOptions={buildOptions(3)}
        introContent="Test intro"
      />
    </ThemeProvider>
  );

  expect(queryByText('Test intro')).not.toBeNull();
});

it('renders additional help when provided', () => {
  const { queryByText } = render(
    <ThemeProvider theme={viaTheme}>
      <RadioGroup
        name="test"
        radioOptions={buildOptions(3)}
        additionalHelpContent="Test additional help"
      />
    </ThemeProvider>
  );

  expect(queryByText('Test additional help')).not.toBeNull();
});
