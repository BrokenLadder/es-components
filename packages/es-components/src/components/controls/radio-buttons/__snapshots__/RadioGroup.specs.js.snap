// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup component renders as expected 1`] = `
<fieldset
  className="es-fieldset sc-dnqmqq eDDNas sc-htoDjs kAbmqn"
>
  <label
    className="es-radio sc-ifAKCX ckpGqX sc-bdVaJa yehor"
    disabled={false}
    htmlFor="test-option-1"
  >
    <input
      checked={true}
      className="es-radio__input--default sc-EHOje iFEyAP"
      disabled={false}
      id="test-option-1"
      name="test"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <span
      className="es-radio__fill sc-bZQynM cnpyBY"
      fill="#069"
    />
    Option 0
  </label>
  <label
    className="es-radio sc-ifAKCX fQsAqZ sc-bdVaJa yehor"
    disabled={false}
    htmlFor="test-option-2"
  >
    <input
      checked={false}
      className="es-radio__input--default sc-EHOje iFEyAP"
      disabled={false}
      id="test-option-2"
      name="test"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <span
      className="es-radio__fill sc-bZQynM cFxFaC"
      fill="#fff"
    />
    Option 1
  </label>
  <label
    className="es-radio sc-ifAKCX fQsAqZ sc-bdVaJa yehor"
    disabled={false}
    htmlFor="test-option-3"
  >
    <input
      checked={false}
      className="es-radio__input--default sc-EHOje iFEyAP"
      disabled={false}
      id="test-option-3"
      name="test"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <span
      className="es-radio__fill sc-bZQynM cFxFaC"
      fill="#fff"
    />
    Option 2
  </label>
</fieldset>
`;

exports[`RadioGroup component renders as expected with legend text and extraContent 1`] = `
<fieldset
  className="es-fieldset sc-dnqmqq eDDNas sc-htoDjs kAbmqn"
>
  <legend
    className="es-fieldset__legend sc-gzVnrw fEWCfG"
  >
    Test legend
  </legend>
  <label
    className="es-radio sc-ifAKCX fQsAqZ sc-bdVaJa yehor"
    disabled={false}
    htmlFor="test-option-1"
  >
    <input
      checked={false}
      className="es-radio__input--default sc-EHOje iFEyAP"
      disabled={false}
      id="test-option-1"
      name="test"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <span
      className="es-radio__fill sc-bZQynM cFxFaC"
      fill="#fff"
    />
    Option 0
  </label>
  <label
    className="es-radio sc-ifAKCX fQsAqZ sc-bdVaJa yehor"
    disabled={false}
    htmlFor="test-option-2"
  >
    <input
      checked={false}
      className="es-radio__input--default sc-EHOje iFEyAP"
      disabled={false}
      id="test-option-2"
      name="test"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <span
      className="es-radio__fill sc-bZQynM cFxFaC"
      fill="#fff"
    />
    Option 1
  </label>
  <label
    className="es-radio sc-ifAKCX fQsAqZ sc-bdVaJa yehor"
    disabled={false}
    htmlFor="test-option-3"
  >
    <input
      checked={false}
      className="es-radio__input--default sc-EHOje iFEyAP"
      disabled={false}
      id="test-option-3"
      name="test"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <span
      className="es-radio__fill sc-bZQynM cFxFaC"
      fill="#fff"
    />
    Option 2
  </label>
</fieldset>
`;
